engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnCompensationSubscriptionBehaviour.java

```diff
@@ -0,0 +1,151 @@
+        context.getTenantId(), context.getProcessInstanceKey());
+  }
+
+  public void activateCompensationHandler(
+      final String elementId, final BpmnElementContext context) {
+
+    // find the boundary event
+    final var activityToCompensate =
+        processState.getFlowElement(
+            context.getProcessDefinitionKey(),
+            context.getTenantId(),
+            BufferUtil.wrapString(elementId),
+            ExecutableActivity.class);
+
+    final ExecutableBoundaryEvent boundaryEvent =
+        activityToCompensate.getBoundaryEvents().stream()
+            .filter(b -> b.getEventType() == BpmnEventType.COMPENSATION)
+            .findFirst()
+            .orElseThrow();
+
+    activateAndCompleteCompensationBoundaryEvent(context, boundaryEvent);
+
+    // activate the compensation handler
+    final ExecutableActivity compensationHandler =
+        boundaryEvent.getCompensation().getCompensationHandler();
+    final DirectBuffer compensationHandlerId = compensationHandler.getId();
+
+    final ProcessInstanceRecord compensationHandlerRecord = new ProcessInstanceRecord();
+    compensationHandlerRecord.wrap(context.getRecordValue());
+    compensationHandlerRecord.setBpmnElementType(compensationHandler.getElementType());
+    compensationHandlerRecord.setElementId(compensationHandlerId);
+
+    commandWriter.appendNewCommand(
+        ProcessInstanceIntent.ACTIVATE_ELEMENT, compensationHandlerRecord);
+  }
+
+  public void createCompensationSubscription(
+      final ExecutableFlowNode element, final BpmnElementContext context) {
+    if (hasCompensationBoundaryEvent(element)) {
+      final var key = keyGenerator.nextKey();
+      final var compensation =
+          new CompensationSubscriptionRecord()
+              .setTenantId(context.getTenantId())
+              .setProcessInstanceKey(context.getProcessInstanceKey())
+              .setProcessDefinitionKey(context.getProcessDefinitionKey())
+              .setCompensableActivityId(BufferUtil.bufferAsString(context.getElementId()))
+              .setCompensableActivityScopeId(context.getFlowScopeKey());
+      stateWriter.appendFollowUpEvent(key, CompensationSubscriptionIntent.CREATED, compensation);
+    }
+  }
+
+  public void updateCompensationSubscription(final BpmnElementContext context) {
+    if (isCompensationThrowEvent(context)) {
+      final var key = keyGenerator.nextKey();
+      final var compensation =
+          new CompensationSubscriptionRecord()
+              .setTenantId(context.getTenantId())
+              .setProcessInstanceKey(context.getProcessInstanceKey())
+              .setProcessDefinitionKey(context.getProcessDefinitionKey())
+              .setThrowEventId(BufferUtil.bufferAsString(context.getElementId()))
+              .setThrowEventInstanceKey(context.getElementInstanceKey());
+      stateWriter.appendFollowUpEvent(key, CompensationSubscriptionIntent.TRIGGERED, compensation);
```

Comment:
:x: We write the `TRIGGERED` event with a new key and not the key of the created subscription. However, `TRIGGERED` is the next state in the lifecycle of the subscription.

We should write the event with the subscription key.

We could retrieve the subscription key and record from the state.

Reply:
❓ We don't have the `compensableActivityId` here, how we can retrieve the compensation subscription ? We should traverse again the process ?

Reply:
Yes, currently the method doesn't contain the compensation activity id. However, the method is called from the processors that know the activity id. We could pass the activity id from the processor in the method. :rocket: 

:wrench: Going further, we could improve the design a bit and follow the information-hiding + separation of concerns principle. The processor retrieves the activity ids, iterates over the ids, and calls `triggerCompensationSubscription()` + `activateCompensationHandler()`. We could combine these calls in the compensation behavior. The processor only needs the information if any compensation handler was invoked or not.

If the compensation state method `findSubscriptionsByProcessInstanceKey()` would return the subscriptions then we don't need to fetch them again in the compensation behavior. We could write the `TRIGGERED` event directly for the subscriptions.